### PR TITLE
Consistent json error

### PR DESCRIPTION
**Pull request by WillAyd**
This will take a few passes to do comprehensively, but trying to clean up error handling in the extension module. Right now failure points are allowed, which can lead to segfaults or surprising behaviour when debugging.

Trying to push closers towards the CPython error handling conventions:

https://docs.python.org/3/c-api/exceptions.html

So basically:

  - Explicitly check for NULL from most C API functions except integer returning functions (where they return -1 and set PyErr) AND
  - Don't set error messages from failing functions, as they should have already set it

### CODE REVIEW

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 1]
```diff
@@ -156,24 +156,72 @@ void *initObjToJSON(void) {
...
 
     mod_pandas = PyImport_ImportModule("pandas");
-    if (mod_pandas) {
-        cls_dataframe =
-            (PyTypeObject *)PyObject_GetAttrString(mod_pandas, "DataFrame");
-        cls_index = (PyTypeObject *)PyObject_GetAttrString(mod_pandas, "Index");
-        cls_series =
-            (PyTypeObject *)PyObject_GetAttrString(mod_pandas, "Series");
-        cls_timedelta = PyObject_GetAttrString(mod_pandas, "Timedelta");
+    if (mod_pandas == NULL) {
```
Comment (jbrockmendel) [inline pandas/_libs/src/ujson/python/objToJSON.c]: this should really never happen right?

Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c]: I would think unlikely. Being pedantic for the sake of consistency

**Review by gfyoung — CHANGES_REQUESTED** [review]
I have to say all this `Py_DECREF`-ing looks quite tedious...is it possible for us to `Py_DECREF` these variables after the checks succeed (e.g. do we need `cls_index` again if we find that it isn't `NULL`) ?

**Comment by WillAyd** [comment]
@gfyoung as an alternate I think could set any of these to NULL and use a goto to Py_XDECREF on these. I'll give that a look over the next few days

**Comment by WillAyd** [comment]
Come back to this one later; clearing queue for now

**Pull request closed without merging**